File: Python3/15_3Sum_2pointers.py
class Solution:
    
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        
        def twoSum(i, j, target: int) -> List[int]:
            res = []
            while i < j:
                if nums[i] + nums[j] == target:
                    res.append([nums[i], nums[j]])
                    i += 1
                    j -= 1
                    while i < j and nums[i] == nums[i-1]:
                        i += 1
                    while i < j and nums[j] == nums[j+1]:
                        j -= 1
                elif nums[i] + nums[j] < target:
                    i += 1
                    while i < j and nums[i] == nums[i-1]:
                        i += 1
                else:
                    j -= 1
                    while i < j and nums[j] == nums[j+1]:
                        j -= 1
            return res
        
        nums.sort()
        res = []
        
        i = 0
        while i < len(nums) - 2:
            two_res = twoSum(i+1, len(nums)-1, -nums[i])
            for l in two_res:
                res.append([nums[i], l[0], l[1]])
            i += 1
            while i < len(nums) - 2 and nums[i] == nums[i-1]:
                i += 1
        
        return res

File: C++/15_3Sum-2pointers.cpp
class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        vector<vector<int>> res;
        int len = nums.size();
        if(len < 3) return res;
        sort(nums.begin(), nums.end());
        for(int i = 0; i < len-2; i++){
            int target = -nums[i];
            int front = i+1;
            int back = len-1;
            while(front < back){
                int sum = nums[front] + nums[back];
                if(sum > target) back--;
                else if(sum < target) front++;
                else{
                    res.push_back({-target, nums[front], nums[back]});
                    while(nums[++front] == nums[front-1] && front < back) ;
                    while(nums[--back] == nums[back+1] && front < back) ;
                }
            }
            while(i+1 < len-2 && nums[i+1] == nums[i]) i++;
        }
        return res;
    }
};